### ISSUE TITLE
Scheduler `CMStochasticIterativeScheduler` `add_noise` bug

### ISSUE THREAD

**Issue by matinmoezzi** [huggingface/diffusers#6117]
### Describe the bug

The `add_noise` function of the consistency model scheduler, `CMStochasticIterativeScheduler`, raises a runtime error in training due to inconsistency between scheduler timestep and timestep in the function argument.
`step_indices = [(schedule_timesteps == t).nonzero().item() for t in timesteps]`
If the timestep given in the function is a random int64 array between 0 and `scheduler.config.num_training_steps`, the above line raises an error. 

### Reproduction

       self.scheduler = CMStochasticIterativeScheduler(num_train_timesteps=200)

        x0 = clean_data
        noise = torch.randn(x0.shape, device=x0.device)

        timesteps = torch.randint(
            0,
            self.scheduler.config.num_train_timesteps,
            (batch_size,),
            device=x0.device,
            dtype=torch.int64,
        )

        noisy_data = self.scheduler.add_noise(x0, noise, timesteps)

       Exception has occurred: RuntimeError       (note: full exception trace is shown but execution is paused at: 
     _run_module_as_main)
      a Tensor with 0 elements cannot be converted to Scalar
     File "/home/admin/miniconda3/envs/dynamicsdiffusion/lib/python3.10/site- 
      packages/diffusers/schedulers/scheduling_consistency_models.py", line 413, in <listcomp>

### Logs

```shell
Exception has occurred: RuntimeError       (note: full exception trace is shown but execution is paused at: _run_module_as_main)
a Tensor with 0 elements cannot be converted to Scalar
  File "/home/admin/miniconda3/envs/dynamicsdiffusion/lib/python3.10/site-packages/diffusers/schedulers/scheduling_consistency_models.py", line 413, in <listcomp>
```

### System Info

Diffusers version 0.24.0, 
Pytorch 2.1.1, 
Python 3.10

### Who can help?

@yiyixuxu @patrickvonplaten

**Comment by sayakpaul** [comment]
Cc: @dg845 as well.

**Comment by dg845** [comment]
When initialized, `CMStochasticIterativeScheduler` will create `timesteps` according to a [Karras sigma schedule](https://github.com/huggingface/diffusers/blob/664e931bcbddd47c21c3d4d547221506f588b9e9/src/diffusers/schedulers/scheduling_consistency_models.py#L235), then [applies noise preconditioning](https://github.com/huggingface/diffusers/blob/664e931bcbddd47c21c3d4d547221506f588b9e9/src/diffusers/schedulers/scheduling_consistency_models.py#L143):

https://github.com/huggingface/diffusers/blob/664e931bcbddd47c21c3d4d547221506f588b9e9/src/diffusers/schedulers/scheduling_consistency_models.py#L90-L92

This follows the [original consistency models paper](https://arxiv.org/pdf/2303.01469.pdf) and the [official implementation](https://github.com/openai/consistency_models/tree/main). In particular, this means that the timesteps on the default timestep schedule will not be integers but rather floats in `[sigma_min, sigma_max]`. (You can override the default timestep schedule with custom `timesteps` by calling `CMStochasticIterativeScheduler.set_timesteps` before calling `CMStochasticIterativeScheduler.add_noise`, but note that this will currently always convert the `timesteps` to Karras sigmas and apply noise preconditioning.)

One workaround could be to sample timestep indices rather than directly sampling timesteps (e.g. consider `timesteps` in the reproduction code above to be indices rather than timesteps), and then recover the timesteps by indexing into `scheduler.timesteps` for use in `CMStochasticIterativeScheduler.add_noise`.

**Comment by matinmoezzi** [comment]
Thank you, @dg845. I got it.
To clarify is this what you mean?
```
         self.scheduler = CMStochasticIterativeScheduler(num_train_timesteps=200)

         x0 = clean_data
         noise = torch.randn(x0.shape, device=x0.device)

         timesteps_idx = torch.randint(
            0,
            self.scheduler.config.num_train_timesteps,
            (batch_size,),
            # device=x0.device,
            dtype=torch.int64,
        )
        timesteps = torch.take(self.scheduler.timesteps, timesteps_idx)

        noisy_data = self.scheduler.add_noise(x0, noise, timesteps)
```
